sql: improve spans for IS NULL constraint

RaduBerinde opened the pull request:
At the level of index selection and span generation, the IS NULL constraint
should behave the same as EQ.

This change fixes incorrect/incomplete handling of IS NULL:
 - it was only generating an "end" constraint; it now generates both
   "start" and "end";
 - it was not handled when finding the "exact prefix" formed by index
    constraints;
 - it was not handled when simplifying the filter.

Fixes #20362.

Release note: Improved handling of IS NULL constraints (enhancement, correctness wasn't affected).

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/20366)
<!-- Reviewable:end -->

knz commented:
The "release note" text must state whether this is a bug fix or a general enhancement. I understand that correctness was not affected, am I correct?

---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/sql/opt_index_selection.go, line 847 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20366#-L-E3HQulTd2sDj0rRGh:-L-E3HQulTd2sDj0rRGi:bcim3k3) ([raw file](https://github.com/cockroachdb/cockroach/blob/d7ae63d6671b9bbbb4819fd51dae6f32ae1bbb97/pkg/sql/opt_index_selection.go#L847)):*
> ```Go
> 		// An IS NULL expression allows us to constrain the start of the range
> 		// to begin at NULL.
> 		if c.Right != tree.DNull {
> ```

How does this code react with IS TRUE? IS FALSE?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20366#-:-L-E2wG4AuBsFQPV63Wf:bkzktxa)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
TFTR! Correct,  updated.

---

Review status: 1 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/sql/opt_index_selection.go, line 847 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20366#-L-E3HQulTd2sDj0rRGh:-L-E4hd59h6TzxWewfJR:b-timeyy) ([raw file](https://github.com/cockroachdb/cockroach/blob/d7ae63d6671b9bbbb4819fd51dae6f32ae1bbb97/pkg/sql/opt_index_selection.go#L847)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

How does this code react with IS TRUE? IS FALSE?
</blockquote></details>

I believe these are normalized to something else, from TestNormalizeExpr:
```
		{`d IS TRUE`, `(d = true) AND (d IS NOT NULL)`},
		{`d IS NOT TRUE`, `(d != true) OR (d IS NULL)`},
```
Not sure why the `d IS NOT NULL` is included..

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20366#-:-L-E4sGoLk77rdyWPYSs:b-3hrsfr)*
<!-- Sent from Reviewable.io -->

knz commented:
I believe a "Release note" text in the commit message is still warranted, that just explains that "queries with IS NULL predicates are handled more efficiently".

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> otherwise.

---

Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/sql/opt_index_selection.go, line 847 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20366#-L-E3HQulTd2sDj0rRGh:-L-EToEAS2lK1R6EnQcl:b9f1dz8) ([raw file](https://github.com/cockroachdb/cockroach/blob/d7ae63d6671b9bbbb4819fd51dae6f32ae1bbb97/pkg/sql/opt_index_selection.go#L847)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I believe these are normalized to something else, from TestNormalizeExpr:
```
		{`d IS TRUE`, `(d = true) AND (d IS NOT NULL)`},
		{`d IS NOT TRUE`, `(d != true) OR (d IS NULL)`},
```
Not sure why the `d IS NOT NULL` is included..

</blockquote></details>

Ah, yes, I had forgotten. thanks for the reminder.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20366#-:-L-ETYAxOcQnVrGPjhAQ:b-va4dmj)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
I keep getting conflicting reports as to where the release notes have to go. I heard putting them in the PR description (which is what I did here; I think this makes it into the merge commit?) or putting them in the commits. Which is it? @jseldess

knz commented:
The PR description does not go to the merge commit. For the release notes to be gatherable from the repository they have to go to the commit message.
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

jseldess commented:
Yep, what @knz said. I'll make https://github.com/cockroachdb/cockroach/blob/master/CONTRIBUTING.md#commit-messages more explicit.

RaduBerinde commented:
Ah, I see, thanks. CC @andreimatei who was under the wrong impression on this.

RaduBerinde commented:
Thanks @jseldess, yes, the "preferably the PR message/merge commit" should be corrected.

jseldess commented:
@RaduBerinde, just opened https://github.com/cockroachdb/cockroach/pull/20388. PTAL.

RaduBerinde committed:
sql: improve spans for IS NULL constraint

At the level of index selection and span generation, the IS NULL constraint
should behave the same as EQ.

This change fixes incorrect/incomplete handling of IS NULL:
 - it was only generating an "end" constraint; it now generates both
   "start" and "end";
 - it was not handled when finding the "exact prefix" formed by index
   constraints;
 - it was not handled when simplifying the filter.

Fixes #20362.

Release note (performance improvement): Improved handling of IS NULL
constraints.

RaduBerinde commented:
I have updated the PR with some test fixes (and rebased to include #20409).

knz commented:
:+1:
